### DISCUSSION TITLE
Issue with Script Engine Initialization in IPED API

### DISCUSSION THREAD

**Discussion by davinunesdasilva** [sepinf-inc/IPED#2144] [Q&A] [answered]
Hello IPED Community,

I'm encountering an issue while attempting to utilize the IPED APIs to extract data from an IPED case. Despite successfully compiling my code, I'm facing an error upon execution:

Cannot invoke "javax.script.ScriptEngine.eval(java.io.Reader)" because "this.engine" is null

Here's a snippet of the relevant code:

import java.io.File;
import java.io.IOException;
import javax.script.ScriptException;
import iped.IPEDSearcher;
import iped.IPEDSource;

public class Main {
    public static void main(String[] args) {
        try {
            IPEDSource ipedSource = new IPEDSource(new File("/mnt/evidences/iped-output/"));
            IPEDSearcher searcher = new IPEDSearcher(ipedSource, "type:jpg");
            var result = searcher.search();
            System.out.println(result.getLength());
        } catch (IOException | ScriptException e) {
            e.printStackTrace();
        }
    }
}
I've ensured that I've included all necessary dependencies in my pom.xml file, and the code compiles without errors. However, during execution, it seems that the script engine fails to initialize properly.

I would greatly appreciate any insights or suggestions on how to troubleshoot and resolve this issue. Thank you in advance for your assistance!

Warm regards,
Davi Nunes da Silva

**Comment by lfcnassif** [comment]
Hi @davinunesdasilva. What IPED version are you using? Versions 4.x need java 11. They won't work with java > 14 because the Nashorn javascript engine was removed from the JDK 15 and later. We plan to upgrade to Java 17 in the near future.

**Reply by davinunesdasilva** [reply]
Thank you for the fast response

I am using version 4.1.5

I am using Java 11 now. 
Although the previous error has been resolved, I am now encountering ClassNotFoundExceptions, specifically:
java.lang.NoClassDefFoundError: org/nustaq/serialization/FSTConfiguration

I'm unsure about which dependencies to include to address this issue. Here are the dependencies listed in my pom.xml:

<dependencies>
    
    <dependency>
        <groupId>iped</groupId>
        <artifactId>iped-api</artifactId>
        <version>4.1.5</version>
    </dependency>
    
    
    <dependency>
        <groupId>org.apache.lucene</groupId>
        <artifactId>lucene-core</artifactId>
        <version>9.2.0</version>
    </dependency>
    
    <dependency>
        <groupId>org.slf4j</groupId>
        <artifactId>slf4j-api</artifactId>
        <version>1.7.25</version>
    </dependency>
    
    <dependency>
        <groupId>commons-logging</groupId>
        <artifactId>commons-logging</artifactId>
        <version>1.3.1</version>
    </dependency>
    
    <dependency>
        <groupId>commons-lang</groupId>
        <artifactId>commons-lang</artifactId>
        <version>2.6</version>
    </dependency>
    
    <dependency>
        <groupId>org.apache.logging.log4j</groupId>
        <artifactId>log4j-api</artifactId>
        <version>2.17.1</version>
    </dependency>
    
    <dependency>
        <groupId>org.bouncycastle</groupId>
        <artifactId>bcutil-jdk15on</artifactId>
        <version>1.69</version>
    </dependency>
    
    <dependency>
        <groupId>org.apache.tika</groupId>
        <artifactId>tika-core</artifactId>
        <version>2.4.0</version>
    </dependency>
    
    <dependency>
        <groupId>org.apache.commons</groupId>
        <artifactId>commons-compress</artifactId>
        <version>1.21</version>
    </dependency>
    
    <dependency>
        <groupId>net.sf.sevenzipjbinding</groupId>
        <artifactId>sevenzipjbinding</artifactId>
        <version>16.02-2.01</version>
    </dependency>
    
    <dependency>
        <groupId>com.google.zxing</groupId>
        <artifactId>core</artifactId>
        <version>3.5.3</version>
    </dependency>
    
    <dependency>
        <groupId>com.google.zxing</groupId>
        <artifactId>javase</artifactId>
        <version>3.5.3</version>
    </dependency>
</dependencies>

Do you have a pom.xml with all dependencies necessary to run that on a linux machine?

**Reply by lfcnassif** [reply]
You can look at the pom.xml files inside each iped maven module. Try to include just iped-engine and iped-utils jars, but you would be better including all iped maven modules and their dependencies, or alternatively all libraries inside iped/lib folder.

**Reply by davinunesdasilva** [reply]
I am getting this error now

Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:315)
	at iped.engine.config.TaskInstallerConfig.loadTasks(TaskInstallerConfig.java:80)
	at iped.engine.config.TaskInstallerConfig.getNewTaskInstances(TaskInstallerConfig.java:45)
	at iped.engine.config.Configuration.loadConfigurables(Configuration.java:212)
	at iped.engine.data.IPEDSource.<init>(IPEDSource.java:183)
	at iped.engine.data.IPEDSource.<init>(IPEDSource.java:154)
	at br.mp.mpf.prba.iped.extractor.IpedExtractor.main(IpedExtractor.java:13)
Caused by: java.lang.NullPointerException
	at iped.localization.Messages.getExternalBundle(Messages.java:40)
	at iped.engine.localization.Messages.getString(Messages.java:25)
	at iped.engine.task.ExportFileTask.<clinit>(ExportFileTask.java:112)

**Reply by lfcnassif** [reply, marked as answer]
Are you running within your IDE? If yes, I suggest you to compile your app using just the IPED core dependencies into your pom.xml and then run your app from command line changing the classpath to point to iped-4.1.5\lib\\*

**Reply by davinunesdasilva** [reply]
When I did that, the Java dependence issues went away but now I have python and C dependency issues. Thank you for your help. I Will try now to solve these non Java dependencies
